### ISSUE TITLE
Firebase auth custom claims cleanup

### ISSUE THREAD

**Issue by appsgenie** [RevenueCat/firestore-revenuecat-purchases#71]
One a subscription expires, should the extension remove the "revenueCatEntitlements" from the firebase auth user's custom claims?

**Comment by jefago** [comment]
Yes, it should.

**Comment by appsgenie** [comment]
hi @jefago  thanks for the reply.
I have cases when the claim is still part of the firebase auth claim.. and Idea where to start debugging or if this is a know issue?

**Comment by appsgenie** [comment]
yes I do have both of those collections (as part of the RC firebase extension). and yes, the main reason I reported this is that I get an EXPIRATION event and the claims are still there. I am not sure how to look at the customers collection since that one always lists the entitlements even for expired customers. I guess the expires_date inside each entitlement would tell me if it's expired or not. but with grace periods etc it gets confusing. so to me, I rely on the events collection. once I get EXPIRATION or other flows that represent expiration then I remove user's entitlements on back-end. But the client having these still in the token is confusing..

**Comment by appsgenie** [comment]
Also, what happens if the user is deleted from firebase auth? It seems that the extension currently throws an exception and I am not sure what the consequences of this are..

**Comment by unxavi** [comment]
Bumping this — running into the same problem on `firestore-revenuecat-purchases@0.1.18` with `SET_CUSTOM_CLAIMS=ENABLED`. Store is RC Web Billing (`store: "RC_BILLING"` on the events).

`getActiveEntitlements` in `functions/src/index.ts` filters entitlements by `expires_date >= now-at-webhook-time`:

```ts
return Object.keys(customerPayload.entitlements).filter((entitlementID) => {
  const expiresDate = customerPayload.entitlements[entitlementID].expires_date;
  return expiresDate === null || moment.utc(expiresDate) >= moment.utc();
});
```

The issue is that `EXPIRATION` events triggered by `BILLING_ERROR` fire BEFORE the entitlement's `expires_date`. Sequence we see when a trial fails to convert (insufficient_funds on the renewal charge):

| Event | `event_timestamp_ms` | `expiration_at_ms` |
|---|---|---|
| `BILLING_ISSUE` | T | T + ~1h |
| `CANCELLATION (BILLING_ERROR)` | T | T + ~1h |
| `EXPIRATION (BILLING_ERROR)` | T | T + ~1h |

At webhook processing time `expires_date > now`, so the entitlement is filtered as "still active" and the claim stays as `["<entitlement_id>"]`. Since there's no further webhook after that, the claim is never cleared and the customer keeps whatever access the claim is gating.

The mirror doc at `rc_customers/{uid}` reflects the cancellation correctly — only the claim is stale.

Same pattern presumably applies to any renewal failure where `event_timestamp_ms < expiration_at_ms`, not just trial conversions, and not just Web Billing.

Is this intentional behavior — i.e., is the claim only meant to reflect the entitlement window from the last successful purchase, and consumers are expected to also check `expires_date` from the mirror doc? Or is it considered a bug that the claim isn't cleared when an `EXPIRATION`/`CANCELLATION` event is received?

**Comment by unxavi** [comment]
@jefago are you guy maintaining this extension or is it better to forget about this and go with my custom integration? I ask because I don't see any anyone replying to any new issue

**Comment by jefago** [comment]
Hi @unxavi , we are discussing internally how we can best address this

**Comment by unxavi** [comment]
@jefago any update?

**Comment by unxavi** [comment]
@jefago something to share at this time?
